vnext/Shared/Modules/BlobModule.cpp

```diff
@@ -0,0 +1,222 @@
+             auto data = part["data"].asString();
+             auto bufferPart = vector<uint8_t>(data.begin(), data.end());
+
+             buffer.reserve(buffer.size() + bufferPart.size());
+             buffer.insert(buffer.end(), bufferPart.begin(), bufferPart.end());
+           } else {
+             if (auto state = weakState.lock()) {
+               if (auto instance = state->Module->getInstance().lock()) {
+                 string message = "Invalid type for blob: " + type.asString();
+                 instance->callJSFunction(
+                     "RCTDeviceEventEmitter", "emit", dynamic::array("blobFailed", std::move(message)));
+               }
+             }
+             return;
+           }
+
+           contentHandler->StoreMessage(std::move(buffer), std::move(blobId));
+         }
+       }},
+
+      {"release",
+       [contentHandler = m_contentHandler](dynamic args) // blobId: string
+       {
+         auto blobId = jsArgAsString(args, 0);
+
+         contentHandler->RemoveMessage(std::move(blobId));
+       }}};
+}
+
+#pragma endregion CxxModule
+
+#pragma endregion BlobModule
+
+#pragma region BlobWebSocketModuleContentHandler
+
+#pragma region IWebSocketModuleContentHandler
+
+void BlobWebSocketModuleContentHandler::ProcessMessage(string &&message, dynamic &params) /*override*/ {
+  params["data"] = std::move(message);
+}
+
+void BlobWebSocketModuleContentHandler::ProcessMessage(vector<uint8_t> &&message, dynamic &params) /*override*/ {
+  auto blob = dynamic::object();
+  blob("offset", 0);
+  blob("size", message.size());
+
+  // substr(1, 36) strips curly braces from a GUID.
+  string blobId = winrt::to_string(winrt::to_hstring(GuidHelper::CreateNewGuid())).substr(1, 36);
+  StoreMessage(std::move(message), std::move(blobId));
+
+  params["data"] = std::move(blob);
+  params["type"] = "blob";
+}
+#pragma endregion IWebSocketModuleContentHandler
+
+void BlobWebSocketModuleContentHandler::Register(int64_t socketID) noexcept {
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  m_socketIDs.insert(socketID);
+}
+
+void BlobWebSocketModuleContentHandler::Unregister(int64_t socketID) noexcept {
+  lock_guard<mutex> lock{m_socketIDsMutex};
+  if (m_socketIDs.find(socketID) != m_socketIDs.end())
+    m_socketIDs.erase(socketID);
+}
+
+winrt::array_view<uint8_t>
+BlobWebSocketModuleContentHandler::ResolveMessage(string &&blobId, int64_t offset, int64_t size) noexcept {
+  lock_guard<mutex> lock{m_blobsMutex};
+
+  auto &data = m_blobs.at(std::move(blobId));
+
+  return winrt::array_view<uint8_t>{data};
+}
+
+void BlobWebSocketModuleContentHandler::RemoveMessage(string &&blobId) noexcept {
+  lock_guard<mutex> lock{m_blobsMutex};
+
+  m_blobs.erase(std::move(blobId));
+}
+
+void BlobWebSocketModuleContentHandler::StoreMessage(vector<uint8_t> &&message, string &&blobId) noexcept {
+  lock_guard<mutex> lock{m_blobsMutex};
+
+  m_blobs.insert_or_assign(std::move(blobId), std::move(message));
```

Comment:
>m_blobs [](http://example.com/codeflow?start=2&length=7)

If we go with shared_ptr keeping an immutable blob, then code here must create a new shared_ptr to the new blob. All readers of the previous shared_ptr will safely read the old copy.